hil/api.py

```diff
@@ -130,7 +131,8 @@ def project_detach_node(project, node):
     if num_attachments != 0:
         raise errors.BlockedError("Node attached to a network")
     for nic in node.nics:
-        if nic.current_action is not None:
+        if nic.current_action is not None and \
+           nic.current_action.status == 'PENDING':
```

Comment:
We should probably fail if there's an action in an error state too, no?

Reply:
If we fail when the last action errored out, how will we ever retry then? I am imagining people retrying a couple of times before notifying admins that their requests are failing; and once fixed they try again.

Reply:
The scenario I'm worried about is something like this:

1. A user submits a newtorking action before going on their lunch break
2. The networking action errors out.
3. Since it's been like 30 min since the first one, they doesn't bother to check the status before submitting the second.
4. Networking actions have now happened out of order.

(4) could be a potential security concern, if e.g. the two actions were a remove of one node and an add of another, when the two nodes are not supposed to have been able to reach one another.

One approach would be to have the networking daemon retry a couple times itself.

Reply:
Don't you think it should be on the users to ensure that they check the status of their previous operations before queuing another one? specially because with this API they'll be able to check it. 

In the current version, we say that network operations are guaranteed to happen in-order, but what if there's an error in the switch that doesn't crash the network server, in that case the users will be oblivious that something bad happened. For example, the brocade driver just logs in case of failures.

> One approach would be to have the networking daemon retry a couple times itself.

In case it doesn't go through even on multiple retries, what would our strategy be? Stop the network server so we fail-safe?

Reply:
Let's say we stop queuing any actions on a nic if the last one failed. We would need some way to force it through (when we know the problem has been fixed). If we expect this to fail rarely, then I guess we could let an admin force it through (by checking for admin privileges).